In [26]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import warnings
warnings.filterwarnings("ignore")

In [27]:
df = pd.read_csv('../data/usedCarListing.csv')
del df['Unnamed: 0']
print(df.shape)
df.head()

(9990, 19)


year    make                       model                   sub_model  \
0  2017     BMW                      ['X3']               xDrive28i AWD   
1  2018    Ford                   ['F-150']  XLT SuperCrew 5.5' Box 4WD   
2  2018     GMC  ['Savana', 'Cargo', 'Van']        2500 Short Wheelbase   
3  2017  Nissan                   ['Rogue']                      SV AWD   
4  2017  Toyota                    ['RAV4']                     XLE AWD   

           city state  mileage        price          exterior_color  \
0      Marietta    GA   77,549  ['$16,709']  Mineral White Metallic   
1        Deland    FL   11,624  ['$26,985']            Shadow Black   
2        Pelham    AL   23,997  ['$19,990']            Summit White   
3  Palmetto Bay    FL   78,116   ['$9,997']          Magnetic Black   
4   Great Falls    MT  189,335  ['$13,495']             Super White   

      interior_color mpg_city mpg_hwy                          engine  \
0              Mocha       21      28  2.0L Inline-4 Gas Turbocharged   
1  Medium Earth Gray       16      22       2.7L V-6 Gas Turbocharged   
2            Unknown        N     cty                    6.0L V-8 Gas   
3           Charcoal       25      32               2.5L Inline-4 Gas   
4              Black       22      28               2.5L Inline-4 Gas   

  transmission drive_type fuel_type  \
0    Automatic        AWD       Gas   
1    Automatic        4WD       Gas   
2    Automatic        RWD       Gas   
3    Automatic        AWD       Gas   
4    Automatic        AWD       Gas   

                                     popular_feature  \
0                                                 []   
1   ['Cruise Control', 'Bluetooth', 'Backup Camera']   
2                ['Cruise Control', 'Backup Camera']   
3  ['Front Heated Seats', 'Cruise Control', 'Back...   
4                                                 []   

                                     vehicle_history    cpo  
0  ['0', ' reported accidents', 'Personal or Rent...  False  
1  ['0', ' reported accidents', 'Personal or Rent...  False  
2  ['0', ' reported accidents', 'Personal or Rent...  False  
3  ['2', ' reported accidents', 'Fleet or Mixed U...  False  
4  ['0', ' reported accidents', 'Personal or Rent...  False

There are totally 9990 used car listings with 19 columns in the datasets. The main columns include:
1. year: model year 
2. make: automaker
3. model: model of the vehicle
4. sub_model: more details of the vehicle model
5. city: current city where the car is located
6. state: current state where the car is located
7. mileage: mileage of the car
8. price: dealers' resell price
9. exterior_color: color of the car
10. interior_color: inside color of the car
11. mpg_city: vehicle efficiency running in city conditions, miles per gallon
12. mpg_hwy: vehicle efficiency running in highway conditions, miles per gallon
13. engine: engine specification
14. transmission: automatic or manual transmission type
15. drive_type: drive type of the car
16. fuel_type: fuel type of the car, gasolin/diesel/hybrid/electric etc.
17. popular_feature: the popular features that the car contains
18. vehicle_history: a vector of vehicle past history including number of reported accidents, number of previous owners, title, and usage type.
19. cpo: whether the car is certified pre-owned or not.

## 1. Create new features

In [28]:
# split vehicle history into four variables(num_accident, usage_type, title, num_owner)
split_data = df["vehicle_history"].str.split(",")
data = split_data.to_list()
names = ["1", "2", "3", "4", '5', '6']
new_df = pd.DataFrame(data, columns=names)
df['num_accident'] = new_df['1'].apply(lambda x:x[2:-1])
df['usage_type'] = new_df['3'].apply(lambda x:x[2:-1]) 
df['title'] = new_df['4'].apply(lambda x:x[2:-1])
df['num_owner'] = new_df['5'].apply(lambda x:x[2:-2])
# delete the history column
del df['vehicle_history']
df.head()

year    make                       model                   sub_model  \
0  2017     BMW                      ['X3']               xDrive28i AWD   
1  2018    Ford                   ['F-150']  XLT SuperCrew 5.5' Box 4WD   
2  2018     GMC  ['Savana', 'Cargo', 'Van']        2500 Short Wheelbase   
3  2017  Nissan                   ['Rogue']                      SV AWD   
4  2017  Toyota                    ['RAV4']                     XLE AWD   

           city state  mileage        price          exterior_color  \
0      Marietta    GA   77,549  ['$16,709']  Mineral White Metallic   
1        Deland    FL   11,624  ['$26,985']            Shadow Black   
2        Pelham    AL   23,997  ['$19,990']            Summit White   
3  Palmetto Bay    FL   78,116   ['$9,997']          Magnetic Black   
4   Great Falls    MT  189,335  ['$13,495']             Super White   

      interior_color  ...                          engine transmission  \
0              Mocha  ...  2.0L Inline-4 Gas Turbocharged    Automatic   
1  Medium Earth Gray  ...       2.7L V-6 Gas Turbocharged    Automatic   
2            Unknown  ...                    6.0L V-8 Gas    Automatic   
3           Charcoal  ...               2.5L Inline-4 Gas    Automatic   
4              Black  ...               2.5L Inline-4 Gas    Automatic   

  drive_type fuel_type                                    popular_feature  \
0        AWD       Gas                                                 []   
1        4WD       Gas   ['Cruise Control', 'Bluetooth', 'Backup Camera']   
2        RWD       Gas                ['Cruise Control', 'Backup Camera']   
3        AWD       Gas  ['Front Heated Seats', 'Cruise Control', 'Back...   
4        AWD       Gas                                                 []   

     cpo num_accident              usage_type  title num_owner  
0  False            0  Personal or Rental Use  Clean         1  
1  False            0  Personal or Rental Use  Clean         1  
2  False            0  Personal or Rental Use  Clean         1  
3  False            2      Fleet or Mixed Use  Clean         1  
4  False            0  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

By splitting the vehicle_history columns into four features, which can greatly represent the vehicle historical information.
1. num_accident: number of reported accidents of the car
2. usage_type: whether the car is mainly used for fleet/mixed use or personal/rental use.
3. title: whether this car has a clean, salvage or other kinds of titles.
4. num_owner: number of previous owners of the used car.

## Check the unique values of four newly constructed features

In [29]:
df.num_owner.unique(),df.num_accident.unique(),df.title.unique(),df.usage_type.unique()

(array(['1', '2', '5', '3', '4', '0', '6', '8', '7', '9', '13', 'alvage',
        'rame Damage'], dtype=object),
 array(['0', '2', '1', '3', '4', '5'], dtype=object),
 array(['Clean', 'Salvaged', 'Frame Damaged', 'Theft Recovered',
        'Theft Recovere', 'Lemon', 'Frame Damage'], dtype=object),
 array(['Personal or Rental Use', 'Fleet or Mixed Use'], dtype=object))

I found that there are some wrong values or synonyms('Frame Damaged', 'Frame Damage') in num_owner, title features due to scraping or parsing. The next block is to deal with these wrong data.

In [30]:
### Deal with title (combine theft recovered with theft recovere, combine frame damaged with frame damage)
df["title"]= df["title"].replace('Theft Recovere', "Theft Recovered") 
df["title"]= df["title"].replace('Frame Damage', "Frame Damaged") 
print(df.title.unique())

### Deal with num_owner that are strings
df["num_owner"]= df["num_owner"].replace('alvage', np.nan) 
df["num_owner"]= df["num_owner"].replace('rame Damage', np.nan) 
print(df.num_owner.unique())

['Clean' 'Salvaged' 'Frame Damaged' 'Theft Recovered' 'Lemon']
['1' '2' '5' '3' '4' '0' '6' '8' '7' '9' '13' nan]


## 2. Check for missing values

In [31]:
df.isna().sum()

year               0
make               0
model              0
sub_model          0
city               0
state              0
mileage            0
price              0
exterior_color     0
interior_color     0
mpg_city           0
mpg_hwy            0
engine             0
transmission       0
drive_type         0
fuel_type          0
popular_feature    0
cpo                0
num_accident       0
usage_type         0
title              0
num_owner          4
dtype: int64

In [32]:
# num_owern, numeric missing value, take the median to fillna
df['num_owner'].fillna(df['num_owner'].median(), inplace=True)

In [33]:
## look at mpg features
print(df.mpg_city.unique())
print(df.mpg_hwy.unique())

['21' '16' 'N' '25' '22' '24' '31' '17' '18' '49' '26' '28' '23' '15' '12'
 '19' '20' '126' '27' '29' '14' '30' '13' '44' '101' '43' '66' '104' '121'
 '120' '32' '48' '38' '95' '51' '106' '41' '40' '35' '11' '110' '129' '42'
 '85' '10' '34' '122' '33' '53' '36' '39' '128' '60' '91' '54' '124' '37'
 '8' '137']
['28' '22' 'cty' '32' '30' '33' '40' '23' '25' '47' '38' '31' '36' '34'
 '29' '24' '21' '16' '99' '37' '39' '19' '17' '26' '35' '18' '27' '20'
 '44' '93' '41' '66' '91' '103' '92' '42' '45' '15' '81' '48' '43' '102'
 '105' '82' '49' '46' '109' '51' '53' '100' '90' '50' '94' '101' '14' '13'
 '12' '111' '108']


Found that there are 'N' and 'cty' values in the mpg features, which are wrong data that we need to replace with NA.

In [34]:
## convert 'N' and 'cty' to nan.
df["mpg_city"]= df["mpg_city"].replace('N', np.nan) 
df["mpg_hwy"]= df["mpg_hwy"].replace('cty', np.nan) 

In [35]:
# deal with mpg_city, mpg_hwy
# first convert mpg from string to numeric values
df['mpg_city'] = df['mpg_city'].astype(float)
df['mpg_hwy'] = df['mpg_hwy'].astype(float)

## Fill missing values in mpg based on domain knowledge.
Since mpg is strongly correlated with vehicle fuel_type, especially for alternative fuel vehicles (hybrid, electric, CNG, hydrogen), I will use the group mean to fillna for these types of vehicles.

For gasoline and diesel vehicles, by observation, I found the missing mpg values for these types of vehicles are more like gas-gazzlers. Therefore, I decided to use 5% quantiles to fill these missing values to indicate their extreme low mpgs.

In [40]:
# checking the fuel types that contains missing values
df[df['mpg_city'].isna()]['fuel_type'].unique()

array(['Gas', 'Hydrogen', 'Diesel', 'Plug-In', 'CNG', 'Hybrid'],
      dtype=object)

In [41]:
# checking gasoline and diesel vehicles whose mpg are missing
display(df[(df['fuel_type']=='Gas') & (df['mpg_city'].isna())].head())
display(df[(df['fuel_type']=='Diesel') & (df['mpg_city'].isna())].head())

year    make                       model                     sub_model  \
2    2018     GMC  ['Savana', 'Cargo', 'Van']          2500 Short Wheelbase   
34   2017    Ford  ['Super', 'Duty', 'F-250']        XL Crew Cab 8' Bed 4WD   
180  2004  HUMMER                      ['H2']                           SUV   
185  2017    Ford  ['Super', 'Duty', 'F-250']    XLT Crew Cab 6.75' Bed 4WD   
218  2004   Dodge             ['Ram', '2500']  SLT Quad Cab Regular Bed 4WD   

                 city state  mileage        price exterior_color  \
2              Pelham    AL   23,997  ['$19,990']   Summit White   
34          Lexington    NC   94,903  ['$29,987']          White   
180            McCook    IL  149,795   ['$9,950']          White   
185      Huntersville    NC   62,669  ['$34,787']   Shadow Black   
218  Colorado Springs    CO  159,551  ['$10,300']           Gray   

        interior_color  ...        engine  transmission drive_type fuel_type  \
2              Unknown  ...  6.0L V-8 Gas     Automatic        RWD       Gas   
34   Medium Earth Gray  ...  6.2L V-8 Gas     Automatic        4WD       Gas   
180              Wheat  ...  6.0L V-8 Gas     Automatic        4WD       Gas   
185  Medium Earth Gray  ...  6.2L V-8 Gas     Automatic        4WD       Gas   
218              Black  ...  5.7L V-8 Gas     Automatic        4WD       Gas   

                         popular_feature    cpo num_accident  \
2    ['Cruise Control', 'Backup Camera']  False            0   
34                                    []  False            0   
180                                   []  False            2   
185                                   []  False            0   
218                                   []  False            0   

                 usage_type  title num_owner  
2    Personal or Rental Use  Clean         1  
34   Personal or Rental Use  Clean         1  
180  Personal or Rental Use  Clean         1  
185  Personal or Rental Use  Clean         1  
218      Fleet or Mixed Use  Clean         3  

[5 rows x 22 columns]

year           make                             model  \
203  2017           Ford        ['Super', 'Duty', 'F-250']   
269  2019  Mercedes-Benz    ['Sprinter', 'Cab', 'Chassis']   
270  2013  Mercedes-Benz  ['Sprinter', 'Passenger', 'Van']   
292  2017           Ford        ['Super', 'Duty', 'F-350']   
322  2019  Mercedes-Benz    ['Sprinter', 'Cab', 'Chassis']   

                            sub_model            city state  mileage  \
203        Lariat Crew Cab 8' Bed 4WD        Columbus    OH   18,095   
269                       3500XD 170"         Houston    TX       10   
270                2500 High Roof LWB   National City    CA  159,212   
292  Platinum Crew Cab 8' Bed 4WD SRW       Hillsboro    OR   38,755   
322                       3500XD 170"  Machesney Park    IL      300   

           price exterior_color interior_color  ...  \
203  ['$34,999']          White    Unspecified  ...   
269  ['$33,995']          White           Gray  ...   
270  ['$18,995']           Gray          Black  ...   
292  ['$42,995']          White    Unspecified  ...   
322  ['$32,979']     Grey White          Black  ...   

                           engine  transmission drive_type fuel_type  \
203  6.7L V-8 Diesel Turbocharged     Automatic        4WD    Diesel   
269  3.0L V-6 Diesel Turbocharged     Automatic        RWD    Diesel   
270  3.0L V-6 Diesel Turbocharged     Automatic        RWD    Diesel   
292  6.7L V-8 Diesel Turbocharged     Automatic        4WD    Diesel   
322  3.0L V-6 Diesel Turbocharged     Automatic        RWD    Diesel   

    popular_feature    cpo num_accident              usage_type  \
203              []  False            1      Fleet or Mixed Use   
269              []  False            0  Personal or Rental Use   
270              []  False            0      Fleet or Mixed Use   
292              []  False            0  Personal or Rental Use   
322              []  False            0  Personal or Rental Use   

             title num_owner  
203  Frame Damaged         2  
269          Clean         1  
270          Clean         2  
292          Clean         1  
322          Clean         1  

[5 rows x 22 columns]

Diesel and gasoline cars whose mpgs are missing are either large cars or luxury cars, whose mpg will be at the very low level.

## Fillna for mpg features

In [42]:
# use fuel_type (electric, plug-in, hybrid, hydrogen, CNG) to fillna first. 
# deal with mpg_city first.
for i in range(len(df)):
    if (df['fuel_type'][i] == 'Hydrogen' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]= df[df['fuel_type']=='Hydrogen']['mpg_city'].mean()
    elif (df['fuel_type'][i] == 'Plug-In' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Plug-In']['mpg_city'].mean()
    elif (df['fuel_type'][i] == 'Hybrid' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Hybrid']['mpg_city'].mean()
    elif (df['fuel_type'][i] == 'CNG' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='CNG']['mpg_city'].mean()
    # The missing values for Diesel and Gas vehicles are more like gas-gazzlers. So I can't use mean or median
    # to replace the missing values. 
    # I decided to use the 95% quantile values to fillna
    elif (df['fuel_type'][i] == 'Gas' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Gas']['mpg_city'].quantile(q=0.05)
    elif (df['fuel_type'][i] == 'Diesel' and np.isnan(df['mpg_city'][i])):
        df['mpg_city'][i]=df[df['fuel_type']=='Diesel']['mpg_city'].quantile(q=0.05)
    
# then deal with mpg_hwy.      
for i in range(len(df)):
    if (df['fuel_type'][i] == 'Hydrogen' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Hydrogen']['mpg_hwy'].mean()
    elif (df['fuel_type'][i] == 'Plug-In' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Plug-In']['mpg_hwy'].mean()
    elif (df['fuel_type'][i] == 'Hybrid' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Hybrid']['mpg_hwy'].mean()        
    elif (df['fuel_type'][i] == 'CNG' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='CNG']['mpg_hwy'].mean()
    # For diesel vehicles with missing values, I found all diesel vehicles are missing the mpg_hwy. 
    elif (df['fuel_type'][i] == 'Gas' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Gas']['mpg_hwy'].quantile(q=0.05)
    elif (df['fuel_type'][i] == 'Diesel' and np.isnan(df['mpg_hwy'][i])):
        df['mpg_hwy'][i]=df[df['fuel_type']=='Diesel']['mpg_hwy'].quantile(q=0.05)       
        
df.head()

year    make                       model                   sub_model  \
0  2017     BMW                      ['X3']               xDrive28i AWD   
1  2018    Ford                   ['F-150']  XLT SuperCrew 5.5' Box 4WD   
2  2018     GMC  ['Savana', 'Cargo', 'Van']        2500 Short Wheelbase   
3  2017  Nissan                   ['Rogue']                      SV AWD   
4  2017  Toyota                    ['RAV4']                     XLE AWD   

           city state  mileage        price          exterior_color  \
0      Marietta    GA   77,549  ['$16,709']  Mineral White Metallic   
1        Deland    FL   11,624  ['$26,985']            Shadow Black   
2        Pelham    AL   23,997  ['$19,990']            Summit White   
3  Palmetto Bay    FL   78,116   ['$9,997']          Magnetic Black   
4   Great Falls    MT  189,335  ['$13,495']             Super White   

      interior_color  ...                          engine  transmission  \
0              Mocha  ...  2.0L Inline-4 Gas Turbocharged     Automatic   
1  Medium Earth Gray  ...       2.7L V-6 Gas Turbocharged     Automatic   
2            Unknown  ...                    6.0L V-8 Gas     Automatic   
3           Charcoal  ...               2.5L Inline-4 Gas     Automatic   
4              Black  ...               2.5L Inline-4 Gas     Automatic   

  drive_type fuel_type                                    popular_feature  \
0        AWD       Gas                                                 []   
1        4WD       Gas   ['Cruise Control', 'Bluetooth', 'Backup Camera']   
2        RWD       Gas                ['Cruise Control', 'Backup Camera']   
3        AWD       Gas  ['Front Heated Seats', 'Cruise Control', 'Back...   
4        AWD       Gas                                                 []   

     cpo num_accident              usage_type  title num_owner  
0  False            0  Personal or Rental Use  Clean         1  
1  False            0  Personal or Rental Use  Clean         1  
2  False            0  Personal or Rental Use  Clean         1  
3  False            2      Fleet or Mixed Use  Clean         1  
4  False            0  Personal or Rental Use  Clean         1  

[5 rows x 22 columns]

In [43]:
## no more NA for now
df.isna().sum()

year               0
make               0
model              0
sub_model          0
city               0
state              0
mileage            0
price              0
exterior_color     0
interior_color     0
mpg_city           0
mpg_hwy            0
engine             0
transmission       0
drive_type         0
fuel_type          0
popular_feature    0
cpo                0
num_accident       0
usage_type         0
title              0
num_owner          0
dtype: int64

## 3. Convert some categorical variables into numeric
Numeric variable: mileage, price, mpg_city, mpg_hwy, num_accident, num_owner

## deal with mileage for the comma.

In [45]:
df['mileage'] = df['mileage'].apply(lambda x: ''.join(x.split(',')))
df['mileage'] = df['mileage'].astype(int)
df['mileage'].head()

0     77549
1     11624
2     23997
3     78116
4    189335
Name: mileage, dtype: int64

## deal with price

In [46]:
# deal with price
df['price'] = df['price'].apply(lambda x: ''.join(x[3:-2].split(',')))
for i in range(len(df)):
    if len(df['price'][i])==0:
        empty_index = i 
    else:
        df['price'][i]= int(df['price'][i])
df['price'].head()

0    16709
1    26985
2    19990
3     9997
4    13495
Name: price, dtype: object

In [47]:
# drop the rows with missing price information
print(empty_index)
df = df.drop(df.index[empty_index])
print(df.shape)

2332
(9989, 22)


## deal with num_accident and num_owner

In [49]:
df['num_accident'] = df['num_accident'].astype(int)
df['num_owner'] = df['num_owner'].astype(int)

In [50]:
# checking feature types
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9989 entries, 0 to 9989
Data columns (total 22 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   year             9989 non-null   int64  
 1   make             9989 non-null   object 
 2   model            9989 non-null   object 
 3   sub_model        9989 non-null   object 
 4   city             9989 non-null   object 
 5   state            9989 non-null   object 
 6   mileage          9989 non-null   int64  
 7   price            9989 non-null   object 
 8   exterior_color   9989 non-null   object 
 9   interior_color   9989 non-null   object 
 10  mpg_city         9989 non-null   float64
 11  mpg_hwy          9989 non-null   float64
 12  engine           9989 non-null   object 
 13  transmission     9989 non-null   object 
 14  drive_type       9989 non-null   object 
 15  fuel_type        9989 non-null   object 
 16  popular_feature  9989 non-null   object 
 17  cpo           

## 4. Drop duplicate rows

In [51]:
# There are totally 5 duplcated rows
print(df.duplicated().sum())
# drop duplicated rows and keep the first row.
print(df.shape)
df = df.drop_duplicates(keep='first')
print(df.shape)

5
(9989, 22)
(9984, 22)


## 5. Deal with potential outliers/wrong data (mileage, price)

In [21]:
# Check very low mileage and very high mileage
display(df.sort_values(by='mileage',ascending=False).head())
df.sort_values(by='mileage',ascending=True).head()
## The data seems making sense to me. I won't drop any data from mileage here.

year        make                       model  \
3091  2012         Ram                    ['3500']   
6988  2012  Volkswagen                   ['Jetta']   
7744  2006      Toyota                  ['Tundra']   
8430  2001   Chevrolet       ['Silverado', '1500']   
1725  2012        Ford  ['Super', 'Duty', 'F-450']   

                             sub_model         city state  mileage  price  \
3091           SLT Crew Cab 8' Box 2WD    Charlotte    NC   484773  14995   
6988                TDI SportWagen DSG    Kingsport    TN   445612   4900   
7744   SR5 Double Cab V8 RWD Automatic  San Antonio    TX   381822   4499   
8430  LT Extended Cab Standard Box 2WD    Pensacola    FL   366778   3985   
1725      Lariat Crew Cab 172" DRW 4WD      Phoenix    AZ   366633  15888   

       exterior_color                        interior_color  ...  \
3091     Bright White  Dark Slate/Medium Graystone Interior  ...   
6988           Silver                           Titan Black  ...   
7744  Salsa Red Pearl                                 Taupe  ...   
8430           Silver                               Unknown  ...   
1725            White                               Unknown  ...   

                                 engine  transmission drive_type fuel_type  \
3091  6.7L Inline-6 Diesel Turbocharged     Automatic        RWD    Diesel   
6988  2.0L Inline-4 Diesel Turbocharged     Automatic        FWD    Diesel   
7744                       4.7L V-8 Gas     Automatic        RWD       Gas   
8430                       5.3L V-8 Gas     Automatic        RWD       Gas   
1725       6.7L V-8 Diesel Turbocharged     Automatic        4WD    Diesel   

                                popular_feature    cpo num_accident  \
3091  ['Remote Engine Start', 'Cruise Control']  False            1   
6988                                         []  False            1   
7744                                         []  False            0   
8430                                         []  False            0   
1725                                         []  False            0   

                  usage_type  title num_owner  
3091      Fleet or Mixed Use  Clean         2  
6988  Personal or Rental Use  Clean         1  
7744  Personal or Rental Use  Clean         2  
8430  Personal or Rental Use  Clean         2  
1725      Fleet or Mixed Use  Clean         1  

[5 rows x 22 columns]

year           make                           model           sub_model  \
4438  2019          Dodge            ['Grand', 'Caravan']                 SXT   
8562  2019  Mercedes-Benz                     ['A-Class']        A 220 4MATIC   
2043  2019  Mercedes-Benz  ['Sprinter', 'Cab', 'Chassis']           4500 170"   
5304  2020  Mercedes-Benz                     ['C-Class']  C 300 Sedan 4MATIC   
1996  2019  Mercedes-Benz  ['Sprinter', 'Cab', 'Chassis']           4500 170"   

             city state  mileage  price           exterior_color  \
4438      Roanoke    IL        6  21923  White Knuckle Clearcoat   
8562  Lincolnwood    IL        7  33875              Night Black   
2043       Monroe    NC        8  28998                    White   
5304   Lake Bluff    IL        9  39990                    Black   
1996       Monroe    NC       10  28998                    White   

             interior_color  ...                          engine  \
4438  Black/Light Graystone  ...                    3.6L V-6 Gas   
8562                  Black  ...  2.0L Inline-4 Gas Turbocharged   
2043                Unknown  ...    3.0L V-6 Diesel Turbocharged   
5304             Silk Beige  ...  2.0L Inline-4 Gas Turbocharged   
1996                Unknown  ...    3.0L V-6 Diesel Turbocharged   

      transmission drive_type fuel_type  \
4438     Automatic        FWD       Gas   
8562     Automatic        AWD       Gas   
2043     Automatic        RWD    Diesel   
5304     Automatic        AWD       Gas   
1996     Automatic        RWD    Diesel   

                                        popular_feature    cpo num_accident  \
4438  ['Power Trunk/Liftgate', 'Remote Engine Start'...  False            0   
8562                                                 []   True            0   
2043                                                 []  False            0   
5304                                                 []  False            0   
1996                                                 []  False            0   

                  usage_type  title num_owner  
4438  Personal or Rental Use  Clean         0  
8562  Personal or Rental Use  Clean         0  
2043  Personal or Rental Use  Clean         0  
5304  Personal or Rental Use  Clean         0  
1996  Personal or Rental Use  Clean         0  

[5 rows x 22 columns]

By checking the 10 cars with top 5 highest mileage and 5 lowest mileage, I need to check if they are making sense.

For top 5 mileage, the total mileage for 8 years is about 484k and average annua mileage is about 60k, which totally makes sense to me.

For bottom 5 mileage, the mileage is from 6~10, which is a very low number. By checking the model year, num_owner and usage_type, I guess these vehicles are purchased by car rental companies and haven't been driven ever, but due to some known reasons, these cars are listed as used cars online for resell purpose. These data makes sense to me.

Therefore, I keep all the data records at this step.

In [22]:
# Check very low price and very high price
display(df.sort_values(by='price',ascending=False).head(10))
df.sort_values(by='price',ascending=True).head(10)
## The high and low prices seems making sense to me. Luxury used cars have very high price. 
## Old used cars have very low prices
## I won't drop any data from mileage here.

year         make                      model    sub_model        city  \
7204  2015  Rolls-Royce                 ['Wraith']          RWD       Tempe   
7736  2005  Lamborghini             ['Murcielago']        Coupe      Boerne   
5159  2014      Ferrari                     ['FF']    Hatchback  Greensboro   
1690  2010      Ferrari  ['599', 'GTB', 'Fiorano']        Coupe     Ontario   
650   2008      Ferrari  ['599', 'GTB', 'Fiorano']        Coupe      Boerne   
696   2015      Ferrari             ['California']  Convertible   Melbourne   
7698  2008      Ferrari  ['599', 'GTB', 'Fiorano']        Coupe    Marietta   
2501  2012      Ferrari                     ['FF']    Hatchback  Pittsburgh   
5563  2007      Ferrari  ['599', 'GTB', 'Fiorano']        Coupe        Mesa   
4036  2008      Bentley                  ['Azure']  Convertible      Naples   

     state  mileage   price         exterior_color     interior_color  ...  \
7204    AZ    23390  149888                  Black              Black  ...   
7736    TX     4565  148888          Grey Metallic                Red  ...   
5159    NC    15237  144635                   Nero               Nero  ...   
1690    CA    30061  139500  Nero Daytona Metallic               Nero  ...   
650     TX    18075  129888              Corsa Red     Interior Color  ...   
696     FL     5835  128998  Nero Daytona Metallic             Sabbia  ...   
7698    GA    16148  124900       Silverstone Gray     Interior Color  ...   
2501    PA    25400  118075                   Gray                Tan  ...   
5563    AZ    18687  114999          Daytona Black              Black  ...   
4036    FL    10393  109995           Beluga Black  Standard Interior  ...   

                          engine  transmission drive_type fuel_type  \
7204  6.6L V-12 Gas Turbocharged     Automatic        RWD       Gas   
7736               6.2L V-12 Gas     Automatic        AWD       Gas   
5159               6.3L V-12 Gas     Automatic        AWD       Gas   
1690               6.0L V-12 Gas        Manual        RWD       Gas   
650                6.0L V-12 Gas     Automatic        RWD       Gas   
696    3.9L V-8 Gas Turbocharged     Automatic        RWD       Gas   
7698               6.0L V-12 Gas        Manual        RWD       Gas   
2501               6.3L V-12 Gas     Automatic        AWD       Gas   
5563               6.0L V-12 Gas     Automatic        RWD       Gas   
4036   6.8L V-8 Gas Turbocharged     Automatic        RWD       Gas   

     popular_feature    cpo num_accident              usage_type  title  \
7204              []  False            0  Personal or Rental Use  Clean   
7736              []  False            0  Personal or Rental Use  Clean   
5159              []   True            0      Fleet or Mixed Use  Clean   
1690              []  False            0  Personal or Rental Use  Clean   
650               []  False            0  Personal or Rental Use  Clean   
696               []  False            0  Personal or Rental Use  Clean   
7698              []  False            0  Personal or Rental Use  Clean   
2501              []  False            0  Personal or Rental Use  Clean   
5563              []  False            0  Personal or Rental Use  Clean   
4036              []  False            0  Personal or Rental Use  Clean   

     num_owner  
7204         1  
7736         2  
5159         3  
1690         5  
650          5  
696          2  
7698         4  
2501         4  
5563         5  
4036         2  

[10 rows x 22 columns]

year       make                model            sub_model  \
253   2005      Dodge          ['Caravan']        Grand SXT LWB   
710   1999     Suzuki  ['Grand', 'Vitara']     4dr JLX Auto 4WD   
8790  2001       Ford           ['Escape']              XLT FWD   
6055  2000       Ford           ['Escort']            ZX2 Coupe   
6784  2002      Honda            ['Civic']   LX Coupe Automatic   
3623  1998  Chevrolet   ['Monte', 'Carlo']                   LS   
5857  2001      Honda          ['Odyssey']       EX 7-Passenger   
2667  1998      Dodge          ['Caravan']  Base 3-door FWD SWB   
1166  2003      Buick            ['Regal']                   LS   
5777  2002       Ford          ['Mustang']       Standard Coupe   

                  city state  mileage price            exterior_color  \
253            Chicago    IL   183245  1500               Stone White   
710         Manchester    MD   180410  1500                     Black   
8790        Plainfield    IN   254999  1538     Toreador Red Metallic   
6055           Everett    WA   287225  1590     Silver Frost Metallic   
6784           Orlando    FL   245421  1595             Taffeta White   
3623              Kent    WA   188000  1600                 Torch Red   
5857  Bloomfield Hills    MI   233255  1695        Dark Emerald Pearl   
2667              Boaz    AL   215798  1800                      Blue   
1166          Lakewood    NJ   128198  1800  Sterling Silver Metallic   
5777      Happy Valley    OR   233850  1850                     Black   

       interior_color  ...             engine  transmission drive_type  \
253           Unknown  ...       3.8L V-6 Gas     Automatic        FWD   
710              Gray  ...       2.5L V-6 Gas     Automatic        4WD   
8790  Medium Graphite  ...       3.0L V-6 Gas     Automatic        FWD   
6055    Dark Charcoal  ...  2.0L Inline-4 Gas        Manual        FWD   
6784             Gray  ...  1.7L Inline-4 Gas     Automatic        FWD   
3623            Black  ...       3.1L V-6 Gas     Automatic        FWD   
5857            Ivory  ...       3.5L V-6 Gas     Automatic        FWD   
2667          Unknown  ...       3.0L V-6 Gas     Automatic        FWD   
1166      Medium Gray  ...       3.8L V-6 Gas     Automatic        FWD   
5777            Black  ...       3.8L V-6 Gas        Manual        RWD   

     fuel_type     popular_feature    cpo num_accident  \
253        Gas  ['Cruise Control']  False            1   
710        Gas                  []  False            0   
8790       Gas  ['Cruise Control']  False            3   
6055       Gas                  []  False            0   
6784       Gas  ['Cruise Control']  False            0   
3623       Gas                  []  False            0   
5857       Gas  ['Cruise Control']  False            1   
2667       Gas                  []  False            0   
1166       Gas  ['Cruise Control']  False            0   
5777       Gas                  []  False            0   

                  usage_type  title num_owner  
253       Fleet or Mixed Use  Clean         4  
710   Personal or Rental Use  Clean         4  
8790  Personal or Rental Use  Clean         8  
6055  Personal or Rental Use  Clean         3  
6784  Personal or Rental Use  Clean         4  
3623  Personal or Rental Use  Clean         6  
5857  Personal or Rental Use  Clean         1  
2667  Personal or Rental Use  Clean         3  
1166  Personal or Rental Use  Clean         4  
5777  Personal or Rental Use  Clean         2  

[10 rows x 22 columns]

By checking the top 10 and bottom 10 used cars in term of resell price, I found the 10 most expensive cars are all luxury brands such as Rolls-Royce, Ferrari, Bentley and Lamborghini, which makes sense for the extreme high prices.

For the 10 cheapest used cars, they are old cars with very high mileage. Therefore, the low prices make sense here, too.

Therefore, I will also keep all the data from this step.

In [52]:
print(df.shape)

(9984, 22)


In [53]:
df.to_csv('../data/usedCarListingCleaned.csv', encoding = 'utf-8')

After the feature engineering(creating new features and dropping features), dealing with missing values, checking outliers, dropping duplicated rows, there are totally 9984 records and 22 features left.